Feat/feature test

PeaceBreker opened the pull request:
新增service,helper,修改部分Pokemoncontroller邏輯,還有新增所有的測試檔案

Commit:
feat:NaturesTableExistTest

Commit:
feat:NaturesTableColumnExistTest

Commit:
feat:AbilitiesTableExistTest

Commit:
feat:AbilitiesTableColumnExistTest

Commit:
feat:RacesTableExistTest

Commit:
feat:RacesTableColumnExistTest

Commit:
feat:SkillsTableExistTest

Commit:
feat:SkillsTableColumnExistTest

Commit:
feat:SkilltagsTableExistTest

Commit:
feat:SkilltagsTableColumnExistTest

Commit:
feat:PokemonControllerCreateTest

Commit:
feat:LearnSkillLogicTest

Commit:
feat:PokemonService

Commit:
feat:SkillLogic

Commit:
feat:PokemonIndexTest

Commit:
feat:PokemonShowTest

Commit:
feat:PokemonUpdateTest

Commit:
feat:PokemonDestroyTest

Commit:
feat:NatureStoreTest

Commit:
feat:NatureIndexTest

Commit:
feat:NatureUpdateTest

Commit:
feat:AbilityStoreTest

Commit:
feat:AbilityIndexTest

Commit:
feat:AbilityUpdateTest

Commit:
feat:RaceGetRaceByNameTest

Commit:
feat:RaceGetAllRaceTest

Commit:
feat:RaceGetSkillsByRaceIdTest

Commit:
feat:SkillStoreTest

Commit:
feat:SkillIndexTest

Commit:
feat:SkillUpdateTest

Commit:
feat:SkillTest

Jheng-github commented on app/Http/Controllers/PokemonController.php:
```diff
@@ -8,63 +8,41 @@
     public function store(PokemonStoreRequest $request)
     {
+        // PokemonStoreRequest驗證整包$data的初步驗證
         $data = $request->all();
```
$request->all(); 會有一定資安危險 要有這個認知

PeaceBreker replied:
來試試其他用法

Jheng-github commented on app/Http/Controllers/PokemonController.php:
```diff
@@ -97,48 +75,24 @@ public function update(PokemonUpdateRequest $request, $id)
             $skill = $data['skill'];
             $skill = array_map('intval', $skill);
 
-            foreach($skill as $skills){
```
這個變數好像會有點困惑XD...
單個技能跑完迴圈變多個技能

PeaceBreker replied:
這段邏輯主要是是執行判斷，最後只return一個bool值，應該不會有這樣的問題

Jheng-github commented on app/Services/PokemonService.php:
```diff
@@ -0,0 +1,67 @@
+            // 如果是true就把資料原路丟回
+            return true;
+        }   // 如果是false也把資料原路丟回
+        else{return false;}
```
分行

PeaceBreker replied:
感謝學長

Jheng-github reviewed (commented):
～

JYu1999 commented on database/factories/AbilityFactory.php:
```diff
@@ -0,0 +1,26 @@
+     */
+    public function definition(): array
+    {
+        $name = $this->faker->unique()->word;
+        $name = substr($name, 0, 20);
+
```
這裡另外拆出來做的用意是什麼？

PeaceBreker replied:
已修改完成

JYu1999 commented on app/Http/Controllers/PokemonController.php:
```diff
@@ -8,63 +8,41 @@
 
 class PokemonController extends Controller
 {
+    public $learnSkill;
```
learnSkill 是要存什麼？這個變數命名看不出這個變數在幹嘛

PeaceBreker replied:
這是一個依賴注入的方式，先宣告出屬性，之後在下面的建構值call物件，引用裡面的method。

method名稱和屬性命名都有問題，我會在修改一下!!

JYu1999 commented on app/SkillLogic.php:
```diff
@@ -0,0 +1,17 @@
+<?php
+
+
+function SkillLogic($skill, $skillTags)
+{
+    foreach ($skill as $skills) {
+        if (!in_array($skills, $skillTags)) {
+            //判斷是否是可以學習技能，如果不行return false
+            return false;
+        }
+    }
+    //如果可以，return true
+    return true;
+}
+
+
+?>
```
這個檔案是幹嘛的？為什麼需要特地拆一個檔案出來放？

PeaceBreker replied:
這個檔案是自己寫的邏輯，主要為驗證寶可夢是否可以學習這個skill。
由於想練習寫unit test的緣故，所以要將這個邏輯獨立出來測試，因此把他包到helper function中。

JYu1999 commented on app/Http/Controllers/AbilityController.php:
```diff
@@ -10,7 +10,7 @@ class AbilityController extends Controller
+            'name' => 'required|max:20|unique:abilities',
         ]);
         try {
             $Ability = Ability::create([
```
為什麼前面的 ability 要大寫？

PeaceBreker replied:
已修改完成

JYu1999 commented on app/Services/PokemonService.php:
```diff
@@ -0,0 +1,67 @@
+
+class PokemonService
+{
+    public function LearnSkillLogic($evolution)
```
function 命名需調整

PeaceBreker replied:
已修改完成

JYu1999 commented on app/Services/PokemonService.php:
```diff
@@ -0,0 +1,67 @@
+            $raceId = $pokemon['race_id'];
+        }
+        //不確定邏輯是否正確，程式碼保留
+        //$pokemon = Pokemon::find($id);
+        //$targetLevel = $data['level'];
+        //$raceId = $pokemon['race_id'];
```
不需要的程式碼請不要留著，git 就有留紀錄了

PeaceBreker replied:
已修改完成

JYu1999 reviewed (commented):
- 不需要的程式碼和註解，就直接拿掉
- 變數 & 函式命名需特別注意
- 註解盡量避免太口語化
- 請使用約定式提交，我看不太懂每個 commit 在幹嘛 -> 目前還沒有鎖分支，請整理好 commit 再重新上傳

Commit:
feat:新增SkillLogic功能
refactor(pokemoncontroller):簡化控制器邏輯至服務
fix:修正PokemonService邏輯
refactor(learnskilllogictest):改寫邏輯
chore:刪除PokemonEvolutionLogicTest
refactor(composer.json):改寫autoload邏輯

Commit:
feat:新增PokemonService功能
refactor(pokemoncontroller): 簡化控制器邏輯至PokemonService
refactor(test): 編輯 LearnSkillLogicTest
feat:新增測試功能PokemonEvolutionLogicTest

Commit:
feat:新增測試功能SkilltagsTableColumnExistTest

Commit:
feat:新增測試功能RaceGetSkillsByRaceIdTest

Commit:
feat:新增測試功能SkillsTableColumnExistTest

Commit:
feat:新增測試功能SkillsTableExistTest

Commit:
feat:新增測試功能AbilitiesTableExistTest

Commit:
feat:新增測試功能PokemonDestroyTest
chore:調整PokemonShowTest程式碼
chore:調整PokemonUpdateTest程式碼

Commit:
feat:新增測試功能RacesTableColumnExistTest

Commit:
feat:新增測試功能RacesTableExistTest

Commit:
feat:新增測試功能PokemonUpdateTest
refactor:改寫PokemonService部分邏輯
feat: 新增PokemonShowTest程式碼
refactor:改寫PokemonController部分邏輯

Commit:
feat:新增測試功能PokemonShowTest
feat:新增功能PokemonFactory
refactor:修正PokemonControllerCreateTest刪除不必要變數

Commit:
feat:新增測試功能RaceGetRaceByNameTest

Commit:
feat:新增測試功能SkilltagsTableExistTest

Commit:
feat:新增測試功能SkillIndexTest

Commit:
feat:新增測試功能AbilitiesTableColumnExistTest

Commit:
feat:新增測試功能PokemonIndexTest

Commit:
feat:新增測試功能NaturesTableExistTest

Commit:
feat:新增測試功能NaturesTableColumnExistTest

Commit:
feat:新增測試功能AbilityIndexTest
refactor: 重構NatureIndexTest註解不必要的程式碼
refactor: 重構NatureStoreTest註解不必要的程式碼
refactor: 重構AbilityStoreTest刪除不必要的程式碼

Commit:
feat:新增測試功能SkillStoreTest

Commit:
feat:新增測試功能NatureStoreTest
chore:調整PokemonFactory程式碼
fix:修正所有從PokemonFactory拿資料測試失敗的問題

Commit:
feat:新增測試功能PokemonControllerCreateTest
feat:新增AbilityFactory功能
feat:新增NatureFactory功能
feat:新增RaceFactory功能
feat:新增SkillFactory功能
feat:新增SkillTagFactory功能
fix:現有測試功能消除註解

Commit:
feat:新增測試功能NatureIndexTest

Commit:
feat:新增測試功能LearnSkillLogicTest

Commit:
feat:新增測試功能NatureUpdateTest

Commit:
feat:新增測試功能AbilityStoreTest

Commit:
feat:新增測試功能AbilityUpdateTest

Commit:
feat:新增測試功能RaceGetAllRaceTest

Commit:
feat:SkillTest資料夾組織程式碼

Commit:
feat:新增測試功能SkillUpdateTest
refactor: 重構NatureUpdateTest刪除不必要的程式碼

Commit:
refactor:重構PokemonController刪除不必要的程式碼
fix:PokemonService分行

Commit:
refactor:刪除不必要的註解

Commit:
refactor(factory): 簡化程式碼

Commit:
refactor: 重構函數和變數名稱以符合正確的命名慣例

Commit:
refactor:優化程式碼組織(所有檔案縮排)和Controller回傳狀態碼

Commit:
refactor:使用request->validated()替換request->all()

Commit:
refactor:重構controller中的訊息使用配置文件

將controller中的訊息文本重構為使用配置文件中的常數，以便統一管理和容易擴展。
新增了http_error_message和http_success_message作為錯誤和成功消息的配置文件。

Commit:
feat:新增AbilityRequest並簡化AbilityController程式碼

Commit:
feat:新增NatureRequest並簡化NatureController程式碼

Commit:
feat:新增SkillRequest並簡化SkillController程式碼

Commit:
feat:新增UploadRequest並簡化UploadController及UploadTwoController程式碼

Commit:
Merge branch 'chore/code-cleanup' into feat/feature_test

Commit:
Merge branch 'feat/feature_test' of github.com:Goodidea-backend-camp/BE11-pokemon-patrick-backend into feat/feature_test

JYu1999 commented on app/Console/Commands/ImportPokemonData.php:
```diff
@@ -14,17 +14,17 @@ class ImportPokemonData extends Command
-    $startId = 1;
-    $endId = 1000;
+    {
+        $startId = 1;
+        $endId = 1000;
```
數字請改用 const

PeaceBreker replied:
已修改完成

JYu1999 commented on tests/Unit/LearnSkillLogicTest.php:
```diff
@@ -11,13 +11,10 @@ class LearnSkillLogicTest extends TestCase
     
     public function testUserCanLearnSkill(): void
```
這個 function 看起來是要測試使用者能不能學技能？

PeaceBreker replied:
是的

JYu1999 commented on tests/Feature/SkillTest/SkillStoreTest.php:
```diff
@@ -8,9 +8,7 @@
-     * A basic feature test example.
-     */
+    
     public function test_example(): void
```
這個 test_example 是殺毀

PeaceBreker replied:
已修改完成

JYu1999 commented on tests/Feature/SkillTest/SKillUpdateTest.php:
```diff
@@ -9,9 +9,7 @@
-     * A basic feature test example.
-     */
+    
     public function test_example(): void
```
這個 test_example 是殺毀

PeaceBreker replied:
已修改完成

JYu1999 commented on tests/Feature/SkillTest/SKillIndexTest.php:
為什麼 k 要大寫？

PeaceBreker replied:
已修改完成

JYu1999 commented on tests/Feature/SkillTest/SKillUpdateTest.php:
為什麼 K 要大寫？

PeaceBreker replied:
已修改完成

Commit:
chore: 重新檢視變數和函數名稱
feat: 安裝 JWT
